internal/solver/database.go

```diff
@@ -0,0 +1,178 @@
+	if !ok {
+		// TODO what happens if there's no packages that satisfy the version range
+		return fps, weights
+	}
+	for _, fp := range mapOfVersions {
+		fps = append(fps, fp)
+	}
+
+	// Sort fps by weight
+	sort.Slice(fps,
+		func(i, j int) bool {
+			semverI := CalculateSemverDistanceToZero(pkgdb.GetPackageByFingerprint(fps[i]).Version)
+			semverJ := CalculateSemverDistanceToZero(pkgdb.GetPackageByFingerprint(fps[j]).Version)
+			return semverI < semverJ
+		},
+	)
+	// create corresponding weight slice
+	for i := range fps {
+		weights = append(weights, i+1)
+	}
+
+	return fps, weights
+}
+
+func CalculateSemverDistanceToZero(semversion string) (distance int) {
+	sv := semver.MustParse(semversion)
+	distance = (int(sv.Major())*int(math.Pow(10, 13)) + int(sv.Minor())*int(math.Pow(10, 9)) + int(sv.Patch()))
+	return distance
+}
+
+func (pkgdb *PkgDB) DebugPrintDB(logger log.Logger) {
+	logger.Debugf("Printing DB")
+	for _, p := range pkgdb.mapFingerprintToPkg {
+		logger.Debug(p.String())
+	}
+}
+
+func CreatePkgDBInstance() *PkgDB {
+	PkgDBInstance = &PkgDB{
+		mapFingerprintToPkg:          make(map[string]*pkg.Pkg),
+		mapBaseFingerprintToVersions: make(map[string]map[string]string),
+		Mutex:                        &sync.Mutex{},
+	}
+	return PkgDBInstance
+}
+
+func GetPkgdDBInstance() *PkgDB {
+	return PkgDBInstance
+}
+
+// mergePkgs gives you a resulting package that is a copy of the new package,
+// but making sure that unknown inthe known CurrenState and DesiredState is not changed,
+// and only unknown info is getting filled.
+func mergePkgs(old pkg.Pkg, new pkg.Pkg) (result *pkg.Pkg) {
+	result = &old
+
+	// Merge CurrentState and DesiredState
+	// E.g:
+	// PACKAGE                     CurrentState   DesiredState
+	// old (coming from release)   installed      unknown
+	// new (coming from tomodify)  unknown        removed
+	// result                      installed      removed
+	if old.CurrentState == pkg.Unknown {
+		result.CurrentState = new.CurrentState
+	}
+	if old.DesiredState == pkg.Unknown {
+		result.DesiredState = new.DesiredState
+	}
+	if old.PinnedVer == pkg.Unknown {
+		result.PinnedVer = new.PinnedVer
+	}
+
+	// Merge Depends and DependsOptional slices
+	if len(old.DependsRel) == 0 {
+		result.DependsRel = new.DependsRel
+	}
+	if len(old.DependsOptionalRel) == 0 {
+		result.DependsOptionalRel = new.DependsOptionalRel
+	}
+
+	return result
+}
+
+// Add adds a package to the database and returns it's ID. If a package was
+// already present in the database, it makes sure to update it, in a way that
+// only unknown info to that package is added.
+func (pkgdb *PkgDB) Add(p *pkg.Pkg) {
+	pkgdb.Lock()
+	defer pkgdb.Unlock()
+
+	fp := p.GetFingerPrint()
+	pInDB, ok := pkgdb.mapFingerprintToPkg[fp]
+	if ok {
+		// package already in DB, merge
+		pkgdb.mapFingerprintToPkg[p.GetFingerPrint()] =
+			mergePkgs(*pInDB, *p)
+	} else {
+		// package not there, add it
+		pkgdb.mapFingerprintToPkg[fp] = p
+	}
```

Comment:
While I was using a dev/test workflow I ran into a problem. I have a repo with a chart in it. That chart has dependencies. I wanted to test development of those charts without releasing new versions of the charts to the repo. So, I do things like `hypper install ./foo` where `foo` is the directory of the chart. And, foo has a dependency on bar that I'm also developing on. So, I update foo to point to bar in the repo `file://../bar` so it get it from there instead.

This isn't working because the database has foo and bar registered from the charts repo. The fingerprint doesn't take the repo location into account and it appears the instances are merged rather than being separate as they are in different repos.

The instances of charts running in a cluster should not take repos into account. But, when they aren't already running, the chart used needs to be from the right repo location.

Reply:
I have made a series of commits to tackle this successfully.

**First**, I have implemented it so `file://` repositories are relative to the cwd from where Hypper is being executed. This corresponds with commits https://github.com/rancher-sandbox/hypper/pull/176/commits/91838649cd33a8370b81d764e602572ca5b814b3, https://github.com/rancher-sandbox/hypper/pull/176/commits/e10e3822ed0c7916739c663d4d150783abfe2f7d.

E.g: with chart dependencies of demo listed as following:
```yaml
  hypper.cattle.io/shared-dependencies: |
    - name: demo-a
      version: "^0.1.0"
      repository: "file://charts/demo-a"
```
One does: 
```console
~/local-charts/charts/ $ hypper install ./charts/demo --dry-run
```
where the folder structure is `./charts/{demo, demo-a}`.

**Second**, I have implemented Helm's behaviour where `file://` is relative to the parent chart, in commit https://github.com/rancher-sandbox/hypper/pull/176/commits/6f03e33dc2963f13fa28024c124a8f2985dab6aa. This isn't pretty, as in Helm this is only used when building charts with local dependencies inside, but we need to keep an pass the parent chart absolute path around. I have achieved this by saving the parent chart path into a new Package.ParentChartPath field, with all that entails. 
E.g: with chart dependencies of demo listed as following:
```yaml
  hypper.cattle.io/shared-dependencies: |
    - name: demo-a
      version: "^0.1.0"
      repository: "file://../demo-a"
```
One does: 
```console
~/local-charts/charts/ $ hypper install ./whatever-path-to/demo --dry-run
```
where the folder structure is `./whatever-path-to/{demo, demo-a}`.


Apart from that, added tests, fixed installing when no cache/repositories.yaml is present, and made it work for local nested charts.